Simperium/src/main/java/com/simperium/android/CredentialsActivity.java

```diff
@@ -376,31 +381,52 @@ public void onClick(DialogInterface dialog, int which) {
             .show();
     }
 
+    private void showDialogErrorLoginReset() {
+        hideDialogProgress();
+        Context context = new ContextThemeWrapper(CredentialsActivity.this, getTheme());
+        new AlertDialog.Builder(context)
+            .setTitle(R.string.simperium_dialog_title_error)
+            .setMessage(getString(R.string.simperium_dialog_message_login_reset, PASSWORD_LENGTH_MINIMUM))
+            .setNegativeButton(android.R.string.cancel, null)
+            .setPositiveButton(R.string.simperium_button_login_reset,
+                new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(DialogInterface dialog, int which) {
+                        Uri uri = Uri.parse(getString(R.string.simperium_dialog_button_reset_url, getEditTextString(mInputEmail)));
+                        Intent intent = new Intent(Intent.ACTION_VIEW, uri);
+                        startActivity(intent);
+                    }
+                }
+            )
+            .show();
+    }
+
     private void startLogin() {
         final String email = getEditTextString(mInputEmail).trim();
         final String password = getEditTextString(mInputPassword).trim();
 
-        if (isValidPassword(password)) {
+        // Use isValidPasswordLength(false) to check if password meets PASSWORD_LENGTH_MINIMUM.
+        if (isValidPassword(password) && !isEmailPasswordMatch() && isValidPasswordLength(false)) {
```

Comment:
We're now checking `isValidPassword(password) && !isEmailPasswordMatch()` both here and on [line 423](https://github.com/Simperium/simperium-android/pull/219/files#diff-f578252d8746b4d32cf359ef222428b4R423). Three questions about that:

1. We're [trimming the password](https://github.com/Simperium/simperium-android/pull/219/files#diff-f578252d8746b4d32cf359ef222428b4R406), which allows a user to add invalid characters (\n\t) to the end of their password, and a different password (with those invalid characters "trimmed") gets accepted and saved. Similarly, the user could end their password with now-valid space characters, and those spaces would get removed without the user's knowledge. Admittedly, it seems like this kind of works with our current Android implementation because we also trim the password on login, so the user can submit the same password with invalid characters that they signed up with, and it works because we again trim it before sending it to the server. We would have to make sure that trim logic is used on every platform to avoid issues though. It's hard for me to imagine that won't cause us problems at some point.

2. Although the password that gets saved is being trimmed here, we're not trimming it [when we're checking for an email password match inside the `isEmailPasswordMatch` method](https://github.com/Simperium/simperium-android/pull/219/files#diff-f578252d8746b4d32cf359ef222428b4R315-R316). This means that an email/password of "me@gmail.com"/"me@gmail.com  " (note the spaces at the end of the password) is accepted as valid because the untrimmed password does not match the email. But the trimmed password is then the password that gets saved, and the trimmed password is the same as the email, which seems like that is what we're trying to stop with the `isEmailPasswordMatch()` check.

3. What would you think about moving the `!isEmailPasswordMatch()` check inside the `isValidPassword(...)` function? We do the email check always (and only) when we're checking `isValidPassword(...)`, so it would remove some duplication. More importantly, I feel like it makes sense from a logic point of view since the email check is part of checking if a password is valid. I would even consider removing the `isEmailPasswordMatch()` method and just assigning the logic to a boolean variable inside `isValidPassword(...)` because that would make it relatively easy to catch inconsistencies like (2) above.

Reply:
We shouldn't be trimming the `email` and `password` variables anymore since we do allow spaces.  I removed trimming in f63c491.

I moved the logic from `isEmailPasswordMatch` to `isValidPassword` in e29fd60.